### PR TITLE
fix: allowing hostnames in gateway/rpc addr field

### LINKED ISSUE #1374: Error is thrown when providing `--gateway.addr` flag
### Celestia Node version

039facd94104c741f8fa995e80f1321776ae4ad3

### OS

macOS 13.0.1

### Install tools

make go-install

### Others

_No response_

### Steps to reproduce it

1. Init the light node
2. Execute start with gateway flags

```shell
celestia light start --core.ip https://limani.celestia-devops.dev --core.grpc.port 9090 --gateway --gateway.addr localhost --gateway.port 26659           
```

### Expected result

Light Node starts with no error 

### Actual result

Light Node fails with error invalid address format

Note: if tried also with `http://localhost` and other variants -> same outcome with errors

### Relevant log output

```Shell
Error: gateway: invalid listen address format: localhost
```

### Notes

As a mitigation, we can just start with this style to get back the REST API
```shell
celestia light start --core.ip https://limani.celestia-devops.dev --core.grpc.port 9090 --gateway
```

### PR DESCRIPTION
**Pull request by distractedm1nd**
## Overview
Closes #1374 

IP/Hostname sanitization logic has been moved to `utils`. We now accept hostnames as valid addresses for the rpc and gateway, and the core ip parsing logic works the same way as the others now.

## Checklist

- [X] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [X] Visual proof for any user facing features like CLI or documentation updates
- [X] Linked issues closed with keywords

### CODE REVIEW

**Review by Wondertan — DISMISSED** [review]
Great that you extracted those in utils. Bad that we touched on unrelated comments. Can you remove those and extract into a separate PR, so that the squashed commit does not touch unrelated comments?

**Comment by distractedm1nd** [comment]
@Wondertan ya i can remove them. I already opened an issue that will fix the problem of make fmt fixing something the linter doesn't catch

**Review by renaynay — DISMISSED** [review]
beautiful thank u

File: nodebuilder/gateway/config.go [hunk 1]
```diff
@@ -22,10 +23,13 @@ func DefaultConfig() Config {
...
-	if ip := net.ParseIP(cfg.Address); ip == nil {
-		return fmt.Errorf("gateway: invalid listen address format: %s", cfg.Address)
+	sanitizedAddress, err := utils.ValidateAddr(cfg.Address)
+	if err != nil {
+		return fmt.Errorf("gateway: invalid address: %w", err)
 	}
-	_, err := strconv.Atoi(cfg.Port)
+	cfg.Address = sanitizedAddress
+
+	_, err = strconv.Atoi(cfg.Port)
```
Comment (MSevey) [inline nodebuilder/gateway/config.go:32]: nit: is there a reason this wouldn't be part of `ValidateAddr`?

File: nodebuilder/rpc/config.go [hunk 2]
```diff
@@ -20,10 +21,13 @@ func DefaultConfig() Config {
...
-	if ip := net.ParseIP(cfg.Address); ip == nil {
-		return fmt.Errorf("service/rpc: invalid listen address format: %s", cfg.Address)
+	sanitizedAddress, err := utils.ValidateAddr(cfg.Address)
+	if err != nil {
+		return fmt.Errorf("service/rpc: invalid address: %w", err)
 	}
-	_, err := strconv.Atoi(cfg.Port)
+	cfg.Address = sanitizedAddress
+
+	_, err = strconv.Atoi(cfg.Port)
```
Comment (MSevey) [inline nodebuilder/rpc/config.go:30]: feels like it should be, so we don't have to remember to do the extra validation step in multiple places.

File: libs/utils/address_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		})
+	}
+}
+
+func TestValidateAddr(t *testing.T) {
+	var tests = []struct {
+		addr string
+		want string
+	}{
+		// Testcase: ip is valid
```
Comment (MSevey) [inline libs/utils/address_test.go:34]: [non-blocking]
Do we want to support ipv6? Or is that something we've talked about?

File: libs/utils/address_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package utils
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSanitizeAddr(t *testing.T) {
```
Comment (MSevey) [inline libs/utils/address_test.go:9]: Additional Test Cases needed

- https
- no leading protocol
- localhost

File: libs/utils/address_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	for _, tt := range tests {
+		t.Run(tt.addr, func(t *testing.T) {
+			got, err := SanitizeAddr(tt.addr)
+			require.NoError(t, err)
+			require.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestValidateAddr(t *testing.T) {
```
Comment (MSevey) [inline libs/utils/address_test.go:29]: We should add tcp test case.

also could refactor the `tests` since they should be the same for both.

**Review by MSevey — COMMENTED** [review]
Some minor comments, fine as follow ups.

**Review by renaynay — APPROVED** [review]

File: nodebuilder/gateway/config.go (hunk 1 above)
Comment (distractedm1nd) [reply nodebuilder/gateway/config.go:32]: Ref #1385

File: libs/utils/address_test.go (hunk 3 above)
Comment (distractedm1nd) [reply libs/utils/address_test.go:34]: yeah we can add a testcase for it

File: libs/utils/address_test.go (hunk 5 above)
Comment (distractedm1nd) [reply libs/utils/address_test.go:29]: Ref #1385

File: libs/utils/address_test.go (hunk 4 above)
Comment (distractedm1nd) [reply libs/utils/address_test.go:9]: Ref #1385

**Pull request merged**